pkg/syncer/fullsync.go

```diff
@@ -53,7 +54,7 @@ func csiFullSync(ctx context.Context, metadataSyncer *metadataSyncInformer) {
 		var err error
 		if metadataSyncer.coCommonInterface.IsFSSEnabled(ctx, common.CSIMigration) && pv.Spec.VsphereVolume != nil {
 			// For vSphere volumes, the migration service will register volumes in CNS.
-			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, pv.Spec.VsphereVolume.VolumePath)
+			volumeHandle, err = volumeMigrationService.GetVolumeID(ctx, migration.VolumeSpec{VolumePath: pv.Spec.VsphereVolume.VolumePath, StoragePolicyID: pv.Spec.VsphereVolume.StoragePolicyID})
```

Comment:
What if the StoragePolicyID does not exist in SPBM? Would it make sense to ignore the storage policy ID for the CNS register operation?

I have the same questions for similar changes in this PR.

Reply:
We get this StoragePolicyID from the PV spec.

```
"vsphereVolume": {
    "fsType": "ext4",
    "storagePolicyID": "aa6d5a82-1c88-45da-85d3-3d74b91a5bad",
     "storagePolicyName": "vSAN Default Storage Policy",
     "volumePath": "[vsanDatastore] 56631b5f-627a-7a01-5b88-02009d31bab6/kubernetes-dynamic-pvc-b1792c12-35f1-4cfd-946c-7ef84643cdd4.vmdk"
}
```

and we expect this policy ID to remain present in the vSphere, as Volumes are already using this policy. 
This PV spec is also immutable, so we can not edit PV and specify a new policy or remove policy association.

We should not give the flexibility to the user to delete the SPBM policy when it is in use by PVs. 

If we want to make this optional then, when the policy is not present in the vCenter, sever throws a genetic CNS fault and we will need to catch this fault, parse the localized message and make a decision to recall create volume without supplying policy id, which is not optimal.

Reply:
> and we expect this policy ID to remain present in the vSphere, as Volumes are already using this policy.

SPBM remembers the policy association of disks that are attached. Once the disk is detached, SPBM does not know about the policy and will allow policy deletion. If the policy is deleted in SPBM, then this code change affects the metadata sync and full sync workflows.

Reply:
So this is like the requirement of not rename datastore or alter VC inventory, as VCP is very sensitive to such changes.
SPBM policy also should not be deleted by storage admin, and if it is deleted, the volume can not be migrated.

if we have dedicated fault from CNS API to tell that specified policy is not present in VC, we can consider this requirement. Comparing localized messages in the fault will not be good.

Reply:
I'm not suggesting to compare error messages. We can invoke SPBM API to know if the policy exists. Any reason why not to do this check? If the policy exists then the policy will be applied for volume registration, if not then don't apply the policy. Wouldn't this fix the problem I raised?

Reply:
I can add that extra VC call using PBM client that will satisfy this requirement. 

but one more thing to confirm - If VC admin has accidentally deleted SPBM policy but later with VC backup/restore if policy appears back in the VC, we can not apply back restored policy on the volume, without de-registering and re-registering back.

Re-registration of volume with CNS does not consider the policy parameters if the volume is already registered with CNS.

Reply:
As we discussed, lets use the "storagePolicyName" to load the policy Id and use this to register the vmdk as a CNS volume. If the policy is deleted in vCenter, the syncer will fail until the vSphere admin recreates the policy with the same name.

Reply:
done. Addressed as part of - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/287/commits/5a0251a40550d0cc1141b86f040084a2353c80f4